pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -479,17 +504,22 @@ public synchronized boolean isTxnAborted(TxnID txnID, PositionImpl readPosition)
         return snapshotAbortedTxnProcessor.checkAbortedTransaction(txnID);
     }
 
+    /**
+     * Sync max read position for normal publish.
+     * @param position {@link PositionImpl} the position to sync.
+     * @param isMarkerMessage whether the message is marker message, in such case, we
+     *                       don't need to trigger the callback to update lastMaxReadPositionMovedForwardTimestamp.
+     */
     @Override
-    public void syncMaxReadPositionForNormalPublish(PositionImpl position) {
+    public void syncMaxReadPositionForNormalPublish(PositionImpl position, boolean isMarkerMessage) {
         // when ongoing transaction is empty, proved that lastAddConfirm is can read max position, because callback
         // thread is the same tread, in this time the lastAddConfirm don't content transaction message.
         synchronized (TopicTransactionBuffer.this) {
             if (checkIfNoSnapshot()) {
-                this.maxReadPosition = position;
+                updateMaxReadPosition(position, isMarkerMessage);
             } else if (checkIfReady()) {
                 if (ongoingTxns.isEmpty()) {
-                    maxReadPosition = position;
-                    changeMaxReadPositionAndAddAbortTimes.incrementAndGet();
+                    updateMaxReadPosition(position, isMarkerMessage);
```

Comment:
This change means only there are no ongoing TXNs, the variable `lastMaxReadPositionMovedForwardTimestamp` will be changed, which is wrong. We should only call `MaxReadPositionCallBack` after committing/aborting transactions.

Reply:
`syncMaxReadPositionForNormalPublish` is used to update the max read position for non-transactional message production. If there is any ongoning txn, we can't move the max read position. We should call `MaxReadPositionCallBack` when the max read position does change.
I don't get your point, could you provide a more detailed explanation?

Reply:
1. Start TXN `1`
2. Start TXN `2`
3, Commit TXN `1`
4. TXN `2` is stuck.

We need to update `lastMaxReadPositionMovedForwardTimestamp` at step `3`, right? But in your implementation, it will not because there is one ongoing TXN(`2`).

Reply:
At step 3, we will update `lastMaxReadPositionMovedForwardTimestamp` with following logic chain:
`org.apache.pulsar.broker.transaction.buffer.impl.TopicTransactionBuffer#commitTxn` -> `org.apache.pulsar.broker.transaction.buffer.impl.TopicTransactionBuffer#removeTxnAndUpdateMaxReadPosition` -> `org.apache.pulsar.broker.transaction.buffer.impl.TopicTransactionBuffer#updateMaxReadPosition`

Reply:
Ah, sorry, I mistakenly thought `disableCallback` was `doCallback`